### ISSUE TITLE
Crash on unsubscribing a Watcher

### ISSUE THREAD

**Issue by AlexHmelevski** [awslabs/aws-mobile-appsync-sdk-ios#9]
When you have several watchers for the one topic the app will crash trying to cancel the subscription

The problem lies in the line of the code that modifies array while iterating on it regarding the upper loop. Here is the case: 
 - we have 3 watchers for the same topic.
-  for i == 0 we remove one subscription. Right now array has 2 elements
-  However the Loop is from 0 to 3 and it crashes on the i == 2

``` swift 
        for i in 0..<self.topicSubscribersDictionary[topic]!.count {
                        if watcherToRemove.getIdentifier() == subscription.getIdentifier() {
                            // remove that watcher for no further notification
                            self.topicSubscribersDictionary[topic]!.remove(at: i)
                        }
        }
```

In addition I have a concern: as long as there is no enforcing of a single thread and self.topicSubscribersDictionary can be modified (until it's really ensured not)  from the background thread thus making the whole code of stopSubscription to crash due to force unwrapping in many places.

P.s. I made a PR just to provide a solution example

**Comment by rohandubal** [comment]
Hello @AlexHmelevski 

We did a release on Thursday(2.6.14) which fixes this issue. Are you still seeing this issue after updating? If yes, could you please re-base your PR with the latest commit?

Thanks,
Rohan

**Comment by AlexHmelevskiAG** [comment]
Hello @rohandubal . Yes, unfortunately the issue is still there :(

**Comment by rohandubal** [comment]
Hello @AlexHmelevski 

This change is merged with SDK release 2.6.15 and merged. You should be able to pull in the changes through Cocoapods/ Carthage.

Thanks,
Rohan
